### ISSUE TITLE
startup: Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed 

### ISSUE THREAD

**Issue by pecastro** [beakerbrowser/beaker#1338]
Operation System: Fedora release 28
Beaker version: d1243594f87979b1f91212df257a4dc8bccc4057
```$ npm run watch

> beakerbrowser@ watch /download/beaker
> gulp start-watch

[18:57:52] Using gulpfile /download/beaker/gulpfile.js
[18:57:52] Starting 'start-watch'...
[18:57:52] Starting 'watch'...
[18:57:52] Starting 'build'...
[18:57:52] Starting 'bundle'...
[18:57:52] Starting 'burnthemall-maybe'...
[18:57:52] Starting 'burnthemallMaybeTask'...
[18:57:52] Finished 'burnthemallMaybeTask' after 1.17 ms
[18:57:52] Finished 'burnthemall-maybe' after 1.67 ms
[18:57:52] Starting 'bundleTask'...
'@beaker/core/webview' is imported by app/webview-preload.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/strings' is imported by app/builtin-pages/views/library-view.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/error-page' is imported by app/shell-window/pages.js, but could not be resolved – treating it as an external dependency
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'@beaker/core/lib/const' is imported by app/builtin-pages/com/files-browser/files-flat-view.js, but could not be resolved – treating it as an external dependency
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'@beaker/core/lib/strings' is imported by app/background-process/ui/permissions.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/const' is imported by app/background-process/browser.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/const' is imported by app/background-process/analytics.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/error-page' is imported by app/background-process/protocols/beaker.js, but could not be resolved – treating it as an external dependency
'default' is imported from external module 'os' but never used
[18:57:56] Finished 'bundleTask' after 3.62 s
[18:57:56] Finished 'bundle' after 3.62 s
[18:57:56] Starting 'less'...
[18:57:56] Starting 'burnthemall-maybe'...
[18:57:56] Starting 'burnthemallMaybeTask'...
[18:57:56] Finished 'burnthemallMaybeTask' after 410 μs
[18:57:56] Finished 'burnthemall-maybe' after 594 μs
[18:57:56] Starting 'lessTask'...
[18:57:56] Finished 'lessTask' after 11 ms
[18:57:56] Finished 'less' after 12 ms
[18:57:56] Finished 'build' after 3.64 s
[18:57:56] Starting '<anonymous>'...
[18:57:56] Finished '<anonymous>' after 18 ms
[18:57:56] Finished 'watch' after 3.66 s
[18:57:56] Starting '<anonymous>'...
Spawning electron /download/beaker/node_modules/electron/dist/electron
Starting external process dat-daemon
Logfile: /home/pecastro/.config/Beaker Browser/debug.log
dat-daemon ready
Detected Linux. Setting up spell check for Linux.
configureFolderToArchiveWatcher() false false
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
```

**Comment by pfrazee** [comment]
Whoop, that's not good. That means the dat daemon is crashing or failing to load. It's strange that no error is being logged, though.

Quick sanity check, could you run `npm run burnthemall` for me?

**Comment by pecastro** [comment]
I see the same thing as in #1336 during module install.
`npm run watch` throws exactly the same errors.

**Comment by pfrazee** [comment]
@pecastro okay interesting, try this:

```
cd $BEAKER_CLONE_DIR
cd app/node_modules/sodium-native
npm run fetch-libsodium
npm install
```

Then try starting beaker again

**Comment by pecastro** [comment]
```
$ npm run fetch-libsodium

> sodium-native@2.2.6 fetch-libsodium /download/beaker/app/node_modules/sodium-native
> git submodule update --recursive --init

$ npm install

> sodium-native@2.2.6 install /download/beaker/app/node_modules/sodium-native
> node-gyp-build "node preinstall.js" "node postinstall.js"

npm notice created a lockfile as package-lock.json. You should commit this file.
added 309 packages from 169 contributors and audited 702 packages in 7.942s
found 0 vulnerabilities

$ cd ..
$ cd ..
$ cd ..
$ npm run watch

> beakerbrowser@ watch /download/beaker
> gulp start-watch

[22:01:58] Using gulpfile /download/beaker/gulpfile.js
[22:01:58] Starting 'start-watch'...
[22:01:58] Starting 'watch'...
[22:01:58] Starting 'build'...
[22:01:58] Starting 'bundle'...
[22:01:58] Starting 'burnthemall-maybe'...
[22:01:58] Starting 'burnthemallMaybeTask'...
[22:01:58] Finished 'burnthemallMaybeTask' after 1.13 ms
[22:01:58] Finished 'burnthemall-maybe' after 1.69 ms
[22:01:58] Starting 'bundleTask'...
'@beaker/core/webview' is imported by app/webview-preload.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/strings' is imported by app/builtin-pages/views/library-view.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/error-page' is imported by app/shell-window/pages.js, but could not be resolved – treating it as an external dependency
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'@beaker/core/lib/const' is imported by app/builtin-pages/com/files-browser/files-flat-view.js, but could not be resolved – treating it as an external dependency
'TimeoutError' is imported from external module 'beaker-error-constants' but never used
'@beaker/core/lib/strings' is imported by app/background-process/ui/permissions.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/const' is imported by app/background-process/browser.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/const' is imported by app/background-process/analytics.js, but could not be resolved – treating it as an external dependency
'@beaker/core/lib/error-page' is imported by app/background-process/protocols/beaker.js, but could not be resolved – treating it as an external dependency
'default' is imported from external module 'os' but never used
[22:02:01] Finished 'bundleTask' after 3.66 s
[22:02:01] Finished 'bundle' after 3.66 s
[22:02:01] Starting 'less'...
[22:02:01] Starting 'burnthemall-maybe'...
[22:02:01] Starting 'burnthemallMaybeTask'...
[22:02:01] Finished 'burnthemallMaybeTask' after 382 μs
[22:02:01] Finished 'burnthemall-maybe' after 562 μs
[22:02:01] Starting 'lessTask'...
[22:02:01] Finished 'lessTask' after 11 ms
[22:02:01] Finished 'less' after 12 ms
[22:02:01] Finished 'build' after 3.67 s
[22:02:01] Starting '<anonymous>'...
[22:02:01] Finished '<anonymous>' after 17 ms
[22:02:01] Finished 'watch' after 3.69 s
[22:02:01] Starting '<anonymous>'...
Spawning electron /download/beaker/node_modules/electron/dist/electron
Starting external process dat-daemon
dat-daemon ready
Logfile: /home/pecastro/.config/Beaker Browser/debug.log
Detected Linux. Setting up spell check for Linux.
configureFolderToArchiveWatcher() false false
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
Uncaught exception: Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:628:16)
    at EventEmitter.mockedSender.send.args [as send] (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:14:15)
    at send (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:60:20)
    at Stream.data (/download/beaker/app/node_modules/pauls-electron-rpc/lib/export-api.js:195:22)
    at Stream.emit (events.js:182:13)
    at Stream.write (/download/beaker/app/node_modules/emit-stream/index.js:14:18)
    at Stream.stream.write (/download/beaker/app/node_modules/through/index.js:26:11)
    at /download/beaker/app/node_modules/emit-stream/index.js:29:35
    at Array.forEach (<anonymous>)
    at EventEmitter.ev.emit (/download/beaker/app/node_modules/emit-stream/index.js:28:29)
```

**Comment by pfrazee** [comment]
Okay, so it doesn't look like the `sodium-native` issue is causing this.

What does your `app/dat-daemon.js` file look like?

**Comment by pecastro** [comment]
It looks like d7cc8f6e14f5f93ba78e7757558af5d59dc3c58b.
It's a clean checkout of d1243594f87979b1f91212df257a4dc8bccc4057.

**Comment by pfrazee** [comment]
It's strange that we're not getting any errors. I'll make sure it's not happening on my ubuntu but if it's not, I'm unsure I can debug this without more info ☹️

**Comment by pecastro** [comment]
Just let me know what your need. I'll provide if I can ...

**Comment by pfrazee** [comment]
I noticed this issue would occur when a dat daemon process lived *past* a previous Beaker instance, which could happen when a crash occurred. I've updated the daemon to close more reliably.

**Comment by da2x** [comment]
This should be reopened. 

I re-enabled Travis testing on Linux 🐧 (see #1380) and ran head first into this on current master (with [the fix for this](ee2a62f1078e48ae2968c2f7d33f2557943bfac2) included).

**Comment by pfrazee** [comment]
@da2x it's not a fatal error. It just indicates that the daemon tried to send a message to the main process after the main process had closed.
